Spigot-Server-Patches/0622-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch

```diff
@@ -0,0 +1,342 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Yannick Lamprecht <yannicklamprecht@live.de>
+Date: Sun, 20 Dec 2020 14:21:32 +0100
+Subject: [PATCH] Fix hopper lag by making entities look for hoppers.
+
+Every tick hoppers try and find an block-inventory to extract from.
+If no tile entity is above the hopper (which there often isn't) it will
+do a bounding box search for minecart chests and minecart hoppers.
+If it can't find an inventory, it will then look for a dropped item,
+which is another bounding box search.
+This patch eliminates that expensive check by having dropped items and
+minecart hoppers/chests look for hoppers instead.
+Hoppers are tile entities meaning you can do a simple tile entity lookup
+to find the nearest hopper in range.
+Pushing out of hoppers causes a bouding box lookup, which this patch
+replaces with a tile entity lookup.
+
+This patch may causes a decrease in the performance of dropped items,
+which is why it can be disabled in the configuration.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index ae5ed3bd0b663092a4658b24cbd69d37b4e141cb..e868a350c8fb554642b17c2ee02e592a26e8adb0 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -709,4 +709,9 @@ public class PaperWorldConfig {
+     private void fixCuringExploit() {
+         fixCuringZombieVillagerDiscountExploit = getBoolean("game-mechanics.fix-curing-zombie-villager-discount-exploit", fixCuringZombieVillagerDiscountExploit);
+     }
++
++  public boolean isHopperPushBased;
++  private void isHopperPushBased() {
++    isHopperPushBased = getBoolean("hopper.push-based", true);
++  }
+ }
+diff --git a/src/main/java/io/papermc/paper/HopperHelper.java b/src/main/java/io/papermc/paper/HopperHelper.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4b43ccdeb3643bdd7ef331424c5e37683bdc262e
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/HopperHelper.java
+@@ -0,0 +1,35 @@
++package io.papermc.paper;
++
++import net.minecraft.server.Block;
++import net.minecraft.server.BlockChest;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Blocks;
++import net.minecraft.server.IInventory;
++import net.minecraft.server.TileEntity;
++import net.minecraft.server.TileEntityHopper;
++import net.minecraft.server.World;
++
++public class HopperHelper {
++
++        public static TileEntityHopper getHopper(World world, BlockPosition pos) {
```

Comment:
Modified it a little bit. Feel free to request old check like here https://github.com/VictorML11/FluxSpigot/blob/master/PaperSpigot-Server-Patches/0006-Fix-hopper-lag-by-making-entities-look-for-hoppers.patch#L250-L275